docs/concepts/components/children.md

```diff
@@ -0,0 +1,255 @@
+---
+title: Children
+description: 
+---
+
+## General usage
+
+_Most of the time,_ when allowing a component to have children, you don't care 
+what type of children the component has. In such cases, the below example will
+suffice.
+
+```rust
+use yew::prelude::*;
+
+#[derive(Properties, Clone)]
+pub struct ListProps {
+	#[prop_or_default]
+	pub children: Children,
+}
+
+pub struct List {
+	props: ListProps,
+}
+
+impl Component for List {
+	type Properties = ListProps;
+	// ...
+
+	fn view(&self) -> Html {
+		html! {
+			<div class="list">
+				{ for self.props.children.iter() }
+			</div>
+		}
+	}
+}
+```
+
+## Advanced usage
+
+There are some cases where you wouldn't want to use the `Children` type
+and instead use an alternative method:
+
+ - You want to only allow components of a specific type to be used as children for your component.
+
+ - You want to share state between a component and its children.
+
+### Typed children
+
+In cases where you want one type of component to be passed as children to your component,
+you can use `yew::html::ChildrenWithProps<T>`.
+
+```rust
+use yew::prelude::*;
+use yew::html::ChildrenWithProps;
+
+// ...
+
+#[derive(Properties, Clone)]
+pub struct ListProps {
+	#[prop_or_default]
+	pub children: ChildrenWithProps<Item>,
+}
+
+pub struct List {
+	props: ListProps,
+}
+
+impl Component for ListProps {
+	type Properties = ListProps;
+	// ...
+
+	fn view(&self) -> Html {
+		html! {
+			<div class="list">
+				{ for self.props.children.iter() }
+			</div>
+		}
+	}
+}
+```
+
+Of course, sometimes you might need to restrict the children to a few different
+components. In these cases, you have to get a little more hands-on with Yew.
+
+The `derive_more` crate is used here for better ergonomics. If you don't want
+to use it, replace `ItemPropVariants` with this:
+
+```rust
+#[derive(Clone)]
+pub enum ItemPropVariants {
+	MyFirstComponent(MyFirstComponentProps),
+	MySecondComponent(MySecondComponentProps),
+}
+
+impl From<MyFirstComponentProps> for ItemPropVariants {
+	fn from(props: MyFirstComponentProps) -> Self {
+		Self::MyFirstComponent(props)		
+	}
+}
+
+impl From<MySecondComponentProps> for ItemPropVariants {
+	fn from(props: MySecondComponentProps) -> Self {
+		Self::MySecondComponent(props)
+	}
+}
+```
+
+```rust
```

Comment:
One thing that needs to be mentioned for this example is that it doesn't constrain the component type but the properties type instead.
`ChildrenRenderer<Item>` accepts any component that has either `MyFirstComponentProps` or `MySecondComponentProps` as its associated properties type.
This is certainly useful if you want to manipulate the props like in the `Shared properties` section but I feel like we should also add a normal example with only restricts the components type. What do you think?

Reply:
I also feel like this is worth mentioning, but I'm not quite sure where to do it. Should it just be mentioned around line 82?

Reply:
Yeah, that sounds like a good place to mention it.
Maybe we should modify the example to work on the component type instead of the properties? Constraining the properties type is a very specific use case, I don't think we should use it to introduce new users to the concept of typed children.
I do like the "Sharing properties" section though so how about we use it there?

Reply:
What's particularly worrying about the example code right now is that it always renders the properties with either `MyFirstComponent` or `MySecondComponent`. So if we create a new component `MyOtherComponent` which also has `type Properties = MyFirstComponentProps` it won't actually use `MyOtherComponent` in the end, it will replace it `MyFirstComponent`:

```rust
html! {
  <List>
     // this will work but it will be replaced with a `MyFirstComponent` because of the `Into<Html>` implementation
    <MyOtherComponent with MyFirstComponentProps {} />
  </List>
}
```

So working with children that require a specific `Properties` type requires extra care so we don't end up in this situation.

Reply:
:+1: Sounds good, although I'm not sure off the top of my head how you would make it check the component type instead or properties. Might add another PR later on that adds an example for that.

Reply:
That's way easier:
```rust
#[derive(Clone, derive_more::From)]
enum Item {
  First(VChild<MyFirstComponent>),
  Second(VChild<MySecondComponent>),
}
// derive_more takes care of adding `From<VChild<[First | Second]>> for Item`
impl Into<Html> for Item {
  fn into(self) -> Html {
    match self {
      Self::First(child) => child.into(),
      Self::Second(child) => child.into(),
    }
  }
}
```

That's all we need. Now `List` only accepts `MyFirstComponent` or `MySecondComponent`.